mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td

```diff
@@ -147,6 +148,21 @@ class NVVM_SpecialRangeableRegisterOp<string mnemonic, list<Trait> traits = []>
       build($_builder, $_state, resultType, ::mlir::LLVM::ConstantRangeAttr{});
     }]>
   ];
+
+  // Define this method for the InferIntRangeInterface.
+  let extraClassDefinition = [{
+    // Infer the result ranges based on the range attribute.
+    void $cppClass::inferResultRanges(
+        ArrayRef<::mlir::ConstantIntRanges> argRanges,
+        SetIntRangeFn setResultRanges) {
+        if (auto rangeAttr = getOperation()->getAttrOfType<LLVM::ConstantRangeAttr>("range")) {
+          setResultRanges(getResult(), 
+                          {rangeAttr.getLower(), rangeAttr.getUpper(),
+                          rangeAttr.getLower(), rangeAttr.getUpper()});
+        }
+    }
+  }];
```

Comment:
Nit: I would prefer if the implementation would be in cpp and that the tablegen part only contains a simple forwarding to said function. Dealing with C++ code implemented in tablegen can be annoying due to lacking LSP support.

Reply:
I agree of having C++ in general. But in this case, it'll be very verbose. 

`NVVM_SpecialRangeableRegisterOp` is used to generate 33 NVVM OPs. I need to replicate inferResultRanges function for all of them in C++.

Reply:
I meant replacing this TableGen part with a call to an actual implementation function that would be implemented in C++. Considering that this is mainly the declaration itself, it is probably not worth it.

Reply:
Updated the PR that extracts this code to C++